### PR TITLE
Optimize Service Stats

### PR DESCRIPTION
**Pull request by Honcharov12**
All these changes were made for Service Stats' performance improvement.

New version uses functions instead of classes for implementing categorizers, matchers and metrics.
Converts configurations dict to configurations tuple of tuples.
Restores stats collection from tq servers.
Modifies old tests.
Deletes unused modules.

Furthermore, now user should do initialization of all request's fields on their own, not to pass them as arguments to `start_request()` and `finalize()` methods. It also makes Stats Service work faster.

### CODE REVIEW

**Comment by Honcharov12** [comment]
Demo:

Were generated 100000 request by following rule:
```
for i in range(100000):
  if i % 2 == 0:
    app = 'guestbook'
  elif i % 3 == 0:
    app = 'ghostbook'
  else:
    app = 'other'
  
  if i % 4 == 0:
    method = 'GET'
  else:
    method = 'POST'

  if i % 5 == 0:
    resource = '/'
  elif i % 13 == 0:
    resouce = '/tmp'
  else:
    resource = '/home/user'

  if i % 7 == 0:
    status = 400 + i % 3
  elif i % 11 == 0:
    status = 500 + i % 7
  else:
    status = 200 + i % 5
  requests_info.append((app, method, resource, status))
```

Then they were processed.
In current version of Service Stats:
```
class FailedMatcher(matchers.RequestMatcher):
  def matches(self, request_info):
    return 400 <= request_info.status <= 599
FAILED = FailedMatcher()

cumulative_dict  = {
  'all': matchers.ANY,
  '4xx': matchers.CLIENT_ERROR,
  '5xx': matchers.SERVER_ERROR,
  'latency': summarizers.AttrValue('latency'),
  'failed': FAILED,
  categorizers.ExactValueCategorizer('by_app', field='app'): {
    'all': matchers.ANY,
    '4xx': matchers.CLIENT_ERROR,
    '5xx': matchers.SERVER_ERROR,
    'latency': summarizers.AttrValue('latency'),
    'failed': FAILED,
    categorizers.ExactValueCategorizer('by_method', field='method'): {
      'all': matchers.ANY,
      'failed': FAILED,
      'latency': summarizers.AttrValue('latency')
    }
  }
}
stats = stats_manager.ServiceStats('my_service', cumulative_counters=cumulative_dict)

for app, method, resource, status in requests_info:
  req = stats.start_request(app=app, method=method, resource=resource)
  stats.get_recent() # were tests with and without it
  req.finalize(status=status)
```
Without getting recent statistics all requests were processed in 1.365 secs, and with - in 44.714 secs.

After modification
```
FAILED = lambda req_info: 400 <= req_info.status <= 599

cumulative_dict = {
  'all': samples.summarize_any,
  '4xx': samples.summarize_client_error,
  '5xx': samples.summarize_server_error,
  'latency': samples.summarize_latency,
  'failed': FAILED, 
  ('by_app', samples.categorize_by_app): {
    'all': samples.summarize_any,
    '4xx': samples.summarize_client_error,
    '5xx': samples.summarize_server_error,
    'latency': samples.summarize_latency,
    'failed': FAILED,
    ('by_method', samples.categorize_by_method): {
      'all': samples.summarize_any,
      'failed': FAILED,
      'latency': samples.summarize_latency
    }
  }
}

stats = stats_manager.ServiceStats('my_service', cumulative_counters=cumulative_dict)
for app, method, resource, status in requests_info:
  req = stats.start_request()
  req.app = app
  req.method = method
  req.resource = resource
  stats.get_recent()
  req.status = status
  req.finalize()
```
were such results: 
 - without getting recent statistics - 0.561 secs;
 - with getting recent statistics - 12.616 secs.

File: AppTaskQueue/appscale/taskqueue/statistics.py [hunk 1]
```diff
@@ -1,92 +1,94 @@
...
-REST_METHOD_CATEGORIZER = RestMethodCategorizer(
-  categorizer_name="by_rest_method")
-PB_STATUS_CATEGORIZER = PBStatusCategorizer(
-  categorizer_name="by_pb_status")
-REST_STATUS_CATEGORIZER = RestStatusCategorizer(
-  categorizer_name="by_rest_status")
+PB_METHOD_CATEGORIZER = categorize_by_pb_method
+REST_METHOD_CATEGORIZER = categorize_by_rest_method
+PB_STATUS_CATEGORIZER = categorize_by_pb_status
+REST_STATUS_CATEGORIZER = categorize_by_rest_status
```
Comment (jeanleonov) [inline AppTaskQueue/appscale/taskqueue/statistics.py]: The comment is outdated, there's no instantiation any more.
Even more, I think there's no need for these constants any more and you can write function names directly in configuration dict.

File: common/appscale/common/service_stats/samples.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+def summarize_server_error(req_info):
+  """ Can be also used as matcher. """
+  return 500 <= req_info.status <= 599
+
+
+def summarize_latency(req_info):
+  return req_info.latency
+
+
+def count_any(requests):
```
Comment (jeanleonov) [inline common/appscale/common/service_stats/samples.py]: Using "any" is not accurate any more. Previously it was "match any", not is should be "count all". The same for "summarize all".

File: common/appscale/common/service_stats/stats_manager.py [hunk 3]
```diff
@@ -284,38 +275,36 @@ def _finalize_request(self, request_no):
...
   def _increment_counters(self, counters_config, counters_dict, request_info):
-    for counter_pair in iteritems(counters_config):
+    for counter_name, categorizer, summarizer, nested_config in counters_config:
       # Counters config can contain following types of items:
-      #  - str->Matcher
-      #  - Categorizer->Matcher
-      #  - Categorizer->nested config with the same structure
+      #  - str, callable(summarizer)
+      #  - str, callable(categorizer), callable(summarizer)
+      #  - str, callable(categorizer), nested config(tuple)
```
Comment (jeanleonov) [inline common/appscale/common/service_stats/stats_manager.py]: Could you explicitly write `None` where value is missing? e.g.:
```
 - str, None,                  callable(summarizer), None
 - str, callable(categorizer), callable(summarizer), None
 - str, callable(categorizer), None,                 tuple(nested config)
``` 
The same for recent stats.

Comment (jeanleonov) [reply common/appscale/common/service_stats/stats_manager.py]: Or without columns alignment..

**Comment by Honcharov12** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/5852/

**Pull request merged**